### ISSUE TITLE
[rush] rush change --overwrite only effective if run twice in one minute

### ISSUE THREAD

**Issue by mmkal** [microsoft/rushstack#2195]
## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc. ##

**Is this a feature or a bug?**

- [X] Feature
- [x] Bug

(kind of a bit of each - a feature design bug?)

**Please describe the actual behavior.**

```bash
git checkout -b rush-change-repro
# change one of your packages, something like this would probably work in the rushstack repo:
touch apps/rush-lib/src/api/SomeNewFile.ts
git commit -a -m changes

rush change --bulk --message change1 --overwrite --bump-type minor
# ^ creates a change file
git commit -a -m changefile1

rush change --bulk --message change2 --overwrite --bump-type minor
# ^ if you ran it fast enough, overwrites the previous change file
git commit -a -m changefile2

sleep 60
rush change --bulk --message change3 --overwrite --bump-type minor
# ^ creates *another* change file instead of overwriting, because it wasn't run in the same minute as the first call to rush change
git commit -a -m changefile3
```

**What is the expected behavior?**

I don't think there's any buggy code, since it does overwrite a changefile _if_ the generated path name is the same. But it seems to me that the `--overwrite` flag is a bit pointless, given that it almost never actually overwrites anything.

I'd suggest another cli argument for `rush change`, `--suffix`, which allows specifying a custom suffix which can be used instead of the timestamp. It could be passed through to this method, which could fall back to using a timestamp to maintain the current default behaviour: https://github.com/microsoft/rushstack/blob/8b2edd92bbdb496ac3225c5ee132cf064eac9b1f/apps/rush-lib/src/api/ChangeFile.ts#L72-L95

That way, the `--overwrite` would actually be meaningful, because you'd allow overwriting the file `common/changes/my-pkg/my-branch_my-suffix.json`, rather than `common/changes/my-pkg/my-branch_<< rapidly-changing-timestamp >>.json`.

I'd happily put up a PR along these lines, if the team are open to the idea?

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** Rush Multi-Project Build Tool
* **Tool Version:** 5.34.0-pr2172.0
* **Node Version:** 12.18.0
  * **Is this a LTS version?** Maybe? It's what I got from `npm install --global @microsoft/rush`
  * **Have you tested on a LTS version?** See above
* **OS:** Windows 10

**Comment by iclanton** [comment]
Yeah that flag might not actually be too useful. Your assessment of the behavior is correct: the changelog filenames contain the current time with minute granularity.

We could perhaps change the file naming convention to include the current second to avoid this issue. @octogonz, thoughts?

**Comment by mmkal** [comment]
@iclanton the flag _would_ be useful though, if the timestamp suffix could be customised to something explicitly-passed in.

Here's an example PR - in this repo a patch is applied to replace the timestamp with `pr-$PR_NUMBER`. This means that the CI job can push changes to keep the changefile in sync with the PR title: https://github.com/mmkal/ts/pull/171. Note that the first change that was pushed marked the change as "patch", but when the PR was edited, the changefile was overwritten in another commit, marking it as "minor". On a team, this pattern avoids the busywork of needing every developer to run `rush change`, for every pull request. Instead, it encourages meaningful pull request titles. When additional changefiles are needed, it's always still possible to run `rush change` locally anyway.

Changing the timestamp to second granularity wouldn't enable that functionality - and if it came with the removal of the `--overwrite` flag, it would make it even hard to achieve - likely the CI job would need to scan and parse the `common/changes/*` directories, which feels like a leaky abstraction - that's rush's domain.

Would you be open to a PR to see what the `--suffix` change would look like? That way, the repo I linked wouldn't need to patch any code, it could just pass `--suffix "pr-$PR_NUMBER"` in the `rush change` call.

**Comment by octogonz** [comment]
@mmkal Seems like there are actually two bugs here:

1. `rush change` sometimes generates the same filename because `generatePath()` uses timestamps whose granularity is minutes
2. `--overwrite` doesn't do anything meaningful

I have run into #1 myself a few times, when I tried to run `rush change` to add a second change log message, and it instead overwrote the previous message.  But if I waited a minute, then it worked correctly.

If we fix #1 then the design question for `--overwrite` is a little different.  If I added two messages, which message should it overwrite?  The most recently added one?

The proposed `--suffix` CLI parameter seems like a workaround that perhaps would be unnecessary if we improved the design of `rush change`.
